File: tests/browser.py
"""Mobile controls, save lifecycle, version upgrades and offline navigation under a Pages subpath."""
import base64, json, os, socket
from functools import partial
from http.server import ThreadingHTTPServer, SimpleHTTPRequestHandler
from pathlib import Path
from threading import Thread
from playwright.sync_api import sync_playwright, expect
from viewport_probe import probe
ROOT=Path(__file__).resolve().parents[1]
OUT=ROOT/'test-results'; OUT.mkdir(exist_ok=True)
PREFIX='/General-AI-Test-Space-5/'
KEY='deadblock.survival.run.v1'
network=True
legacy=False
class Handler(SimpleHTTPRequestHandler):
 def log_message(self,*args): pass
 def do_GET(self):
  if not network:
   self.connection.shutdown(socket.SHUT_RDWR);self.connection.close();return
  if legacy and self.path.split('?')[0]==PREFIX+'sw.js':
   body=(ROOT/'tests/fixtures/sw-v1.js').read_bytes()
   self.send_response(200);self.send_header('Content-Type','text/javascript');self.end_headers();self.wfile.write(body);return
  super().do_GET()
 def translate_path(self,path):
  if path.startswith(PREFIX):path='/'+path[len(PREFIX):]
  return super().translate_path(path)
server=ThreadingHTTPServer(('127.0.0.1',8765),partial(Handler,directory=str(ROOT)))
Thread(target=server.serve_forever,daemon=True).start()
BASE='http://127.0.0.1:8765'+PREFIX
results=[]
try:
 with sync_playwright() as pw:
  for engine in os.environ.get('BROWSERS','chromium,webkit').split(','):
   opts={'headless':engine!='webkit'}
   if engine=='chromium' and os.environ.get('CHROMIUM_PATH'):opts['executable_path']=os.environ['CHROMIUM_PATH']
   browser=getattr(pw,engine).launch(**opts)
   if engine=='webkit':probe(browser)
   for name,w,h in [('desktop',1365,900),('portrait',390,844),('landscape',844,390),('small-landscape',667,375)]:
    context=browser.new_context(viewport={'width':w,'height':h},has_touch=name!='desktop',is_mobile=name!='desktop')
    page=context.new_page();errors=[];page.on('pageerror',lambda error:errors.append(str(error)))
    try:
     page.goto(BASE+'?test=1');page.bring_to_front();page.wait_for_function('!!window.__deadblock')
     if name=='portrait':
      expect(page.locator('#orientationGate')).to_be_visible();assert page.evaluate('__deadblock.rotationBlocked')
      page.screenshot(path=str(OUT/f'{engine}-portrait-rotate.png'))
      w,h=844,390;page.set_viewport_size({'width':w,'height':h});expect(page.locator('#orientationGate')).to_be_hidden(timeout=10000)
     if name=='desktop':
      sheet=page.evaluate('''()=>{
       const canvas=document.createElement('canvas');
       const r=new __deadblock.renderer.constructor(canvas,{blood:false,shake:false,quality:'high'});
       canvas.width=2240;canvas.height=2520;r.state={weapon:'pistol'};const c=r.c;c.scale(2,2);
       c.fillStyle='#ced0bc';c.fillRect(0,0,1120,1260);c.font='13px monospace';
       const kinds=['player','walker','runner','brute','bomber','cinder','boss'];
       kinds.forEach((kind,row)=>{for(let col=0;col<8;col++){
        const angle=-Math.PI/2+col*Math.PI/4,x=col*140+70,y=row*180+164;
        c.fillStyle='#28362c';c.fillText(kind+' '+(col*45)+'deg',x-55,row*180+19);
        r.actor({kind,x,y,angle,walk:0,fire:0,flash:0,telegraph:0,fuse:-1,hp:100,maxHp:100},kind==='player');
       }});return canvas.toDataURL().split(',')[1];
      }''');(OUT/f'{engine}-model-directions.png').write_bytes(base64.b64decode(sheet))
     page.locator('#newButton').click();assert page.locator('[data-map]').count()==7
     page.screenshot(path=str(OUT/f'{engine}-{name}-maps.png'))
     page.locator('[data-action="begin"]').click()
     page.wait_for_function('!!__deadblock.game && !__deadblock.paused')
     assert page.evaluate('__deadblock.game.s.phase')=='prep'
     assert page.evaluate('__deadblock.game.s.wave')==0
     page.evaluate('__deadblock.game.s.cash=500')
     page.locator('#prepBar [data-action="shop"]').click()
     expect(page.locator('[data-action="buy"][data-id="wall"]')).to_be_visible()
     page.locator('[data-action="buy"][data-id="wall"]').click()
     page.locator('[data-action="prepare"]').click()
     assert page.evaluate('__deadblock.game.s.wave')==0
     page.locator('#weaponDock .more-slot').click();page.locator('[data-action="chooseBuild"][data-id="wall"]').click()
     page.locator('.dialog-foot [data-action="close"]').click()
     page.locator('#placeButton').click()
     assert page.evaluate('__deadblock.game.s.defenses.length')==1
     assert page.evaluate('__deadblock.game.s.elapsed')==0
     expect(page.locator('#grenadeButton')).to_be_disabled()
     page.evaluate('__deadblock.persist()');prep_save=page.evaluate('__deadblock.store.read().data')
     page.reload();page.locator('#continueButton').click();assert page.evaluate('__deadblock.game.s.phase')=='prep'
     assert page.evaluate('__deadblock.game.s.defenses')==prep_save['defenses']
     page.screenshot(path=str(OUT/f'{engine}-{name}-preparation.png'))
     page.locator('#startWaveButton').click();page.wait_for_function('__deadblock.game.s.wave===1',timeout=7000)
     assert page.evaluate('__deadblock.game.s.defenses.length')==1
     page.evaluate('const s=__deadblock.game.s;s.countdown=0;s.player.hurt=10000;s.remaining=1;s.spawnTimer=10000;s.enemies=[]')
     assert page.locator('#autoButton').count()==0 and page.locator('#buildCycle').count()==0
     assert page.locator('#weaponDock button').count()==2
     page.wait_for_timeout(180);assert page.evaluate('__deadblock.game.s.bullets.length')==0
     boxes=page.locator('#loadoutBar button').evaluate_all('(els)=>els.map(e=>{const r=e.getBoundingClientRect();return {x:r.x,y:r.y,w:r.width,h:r.height}})')
     for i,b in enumerate(boxes):
      assert b['x']>=0 and b['x']+b['w']<=w+1 and b['y']>=0 and b['y']+b['h']<=h+1,boxes
      assert b['w']>=44 and b['h']>=44,boxes
      for a in boxes[i+1:]:assert not (min(a['x']+a['w'],b['x']+b['w'])>max(a['x'],b['x'])+.5 and min(a['y']+a['h'],b['y']+b['h'])>max(a['y'],b['y'])+.5),boxes
     x0=page.evaluate('__deadblock.game.s.player.x');move_id=page.evaluate('__deadblock.game.s.id')
     if name=='desktop':
      page.keyboard.down('d');page.keyboard.down(' ');page.wait_for_timeout(250);page.keyboard.up('d');page.keyboard.up(' ')
     else:
      pad=page.locator('#moveStick').bounding_box();a=page.locator('#aimStick').bounding_box()
      if engine=='chromium':
       cdp=context.new_cdp_session(page)
       cdp.send('Input.dispatchTouchEvent',{'type':'touchStart','touchPoints':[{'id':1,'x':pad['x']+pad['width']/2+25,'y':pad['y']+pad['height']/2},{'id':2,'x':a['x']+a['width']/2+28,'y':a['y']+a['height']/2}]})
       page.wait_for_timeout(250)
       assert page.evaluate('__deadblock.currentInput().fire')
       cdp.send('Input.dispatchTouchEvent',{'type':'touchEnd','touchPoints':[]})
      else:
       page.mouse.move(pad['x']+pad['width']/2+25,pad['y']+pad['height']/2);page.mouse.down();page.wait_for_timeout(250);page.mouse.up()
     assert page.evaluate('__deadblock.game.s.player.x')>x0+10, page.evaluate('({x:__deadblock.game.s.player.x,id:__deadblock.game.s.id,paused:__deadblock.paused,countdown:__deadblock.game.s.countdown,elapsed:__deadblock.game.s.elapsed,hidden:document.hidden,visibility:document.visibilityState,renderTime:__deadblock.renderer.time,focus:document.activeElement?.outerHTML,inputs:__deadblock.currentInput()})')
     page.wait_for_timeout(60);assert page.evaluate('__deadblock.currentInput().fire') is False
     page.locator('#weaponDock [data-action="loadout"]').last.click();assert page.evaluate('__deadblock.paused')
     page.locator('[data-action="chooseBuild"][data-id="mine"]').click()
     page.locator('.dialog-foot [data-action="close"]').click();assert page.evaluate('!__deadblock.paused')
     assert 'MINE' in page.locator('#placeLabel').inner_text()
     if name=='landscape':
      before_rotation=page.evaluate('__deadblock.game.s.elapsed')
      page.set_viewport_size({'width':390,'height':844});expect(page.locator('#orientationGate')).to_be_visible()
      paused_time=page.evaluate('__deadblock.game.s.elapsed');page.wait_for_timeout(250)
      assert page.evaluate('__deadblock.game.s.elapsed')==paused_time
      assert page.evaluate('__deadblock.currentInput().fire') is False
      page.set_viewport_size({'width':w,'height':h});expect(page.locator('#orientationGate')).to_be_hidden(timeout=10000)
      expect(page.locator('#panelTitle')).to_have_text('Catch your breath.')
      page.locator('[data-action="resume"]').click()
     page.locator('#pauseButton').click();before=page.evaluate('__deadblock.game.s.elapsed');page.wait_for_timeout(150);assert page.evaluate('__deadblock.game.s.elapsed')==before
     page.locator('[data-action="resume"]').click()
     page.evaluate("__deadblock.game.s.phase='shop';__deadblock.game.s.enemies=[];__deadblock.game.s.remaining=0;__deadblock.game.s.cash=20000;__deadblock.showShop()")
     page.locator('[data-action="tab"][data-id="weapons"]').click()
     for gun in ['smg','shotgun','carbine','flamer','railgun','launcher','minigun']:page.locator(f'[data-action="buy"][data-id="{gun}"]').click()
     assert page.locator('#weaponDock button').count()==2
     page.locator('[data-action="tab"][data-id="upgrades"]').click();page.locator('[data-action="buy"][data-id="power"]').click()
     page.locator('[data-action="tab"][data-id="weapons"]').click();page.screenshot(path=str(OUT/f'{engine}-{name}-shop.png'))
     saved=page.evaluate('__deadblock.store.read().data');page.reload();page.locator('#continueButton').click()
     restored=page.evaluate('__deadblock.game.s')
     for field in ['id','wave','owned','ammo','cash','grenades','upgrades','inventory']:assert saved[field]==restored[field],field
     page.locator('[data-action="prepare"]').click();assert page.evaluate('__deadblock.game.s.phase')=='prep';page.locator('#startWaveButton').click();page.evaluate('__deadblock.game.s.countdown=0;__deadblock.game.s.player.hurt=10000')
     page.wait_for_timeout(200);page.evaluate('__deadblock.game.s.player.hurt=0');page.wait_for_timeout(30);page.screenshot(path=str(OUT/f'{engine}-{name}-game.png'))
     page.evaluate("__deadblock.game.s.player.hurt=0;__deadblock.game.damagePlayer(99999);__deadblock.processEvents()")
     assert page.evaluate('__deadblock.store.read()') is None
     page.get_by_role('button',name='Main menu',exact=True).click();expect(page.locator('#continueButton')).to_be_hidden()
     assert not errors,errors
     results.append({'browser':engine,'layout':name,'passed':True});print(engine,name,'PASS',flush=True)
    except Exception as error:
     metrics=page.evaluate('''()=>({inner:[innerWidth,innerHeight],outer:[outerWidth,outerHeight],client:[document.documentElement.clientWidth,document.documentElement.clientHeight],screen:[screen.width,screen.height],media:matchMedia('(orientation:landscape)').matches,visual:visualViewport?{w:visualViewport.width,h:visualViewport.height,scale:visualViewport.scale}:null,hidden:document.hidden,orientation:screen.orientation?.type,blocked:window.__deadblock?.rotationBlocked,phase:window.__deadblock?.game?.s.phase,wave:window.__deadblock?.game?.s.wave,paused:window.__deadblock?.paused,focus:document.activeElement?.outerHTML,panel:document.querySelector('#panel')?.open})''')
     print('VIEWPORT',page.viewport_size,metrics,flush=True)
     print('FAIL',engine,name,str(error),errors,flush=True);(OUT/f'{engine}-{name}-error.txt').write_text(str(error)+chr(10)+str(errors))
     try:page.screenshot(path=str(OUT/f'{engine}-{name}-FAIL.png'))
     except Exception:pass
     results.append({'browser':engine,'layout':name,'passed':False,'error':str(error),'viewport':metrics})
    finally:context.close()
   # Start with a previous cache version, upgrade using the public landing page,
   # then confirm save restoration with the HTTP server completely unreachable.
   legacy=True;context=browser.new_context();page=context.new_page();page.goto(BASE)
   page.wait_for_function('navigator.serviceWorker.controller!==null')
   page.locator('#newButton').click();page.locator('[data-action="begin"]').click();page.locator('#startWaveButton').click();page.wait_for_timeout(2400)
   page.locator('#pauseButton').click();saved=page.evaluate('(k)=>JSON.parse(localStorage.getItem(k)).data',KEY)
   legacy=False;page.goto(BASE);page.evaluate("""async()=>{
    const reg=await navigator.serviceWorker.getRegistration();await reg.update();
    const deadline=Date.now()+45000;
    while(Date.now()<deadline){
     const keys=await caches.keys();
     if(keys.includes('deadblock-survival-1.3.0')&&!keys.includes('deadblock-survival-1.0.0')&&reg.active?.state==='activated'&&navigator.serviceWorker.controller===reg.active)return true;
     await new Promise(resolve=>setTimeout(resolve,150));
    }
    throw Error('Updated cache exists but its worker did not take control');
   }""");page.goto(BASE+'update.html');page.locator('#install').click();page.wait_for_url('**/?v=1.3.0',timeout=60000)
   page.locator('#continueButton').click();page.locator('#pauseButton').click()
   after=page.evaluate('(k)=>JSON.parse(localStorage.getItem(k)).data',KEY);assert saved['id']==after['id']
   page.locator('[data-action="menu"]').click();network=False;page.reload(wait_until='load')
   page.locator('#continueButton').click();expect(page.locator('#hud')).to_be_visible()
   assert page.evaluate('(k)=>JSON.parse(localStorage.getItem(k)).data.id',KEY)==saved['id']
   network=True;context.close();browser.close()
   results.append({'browser':engine,'layout':'cache-update-and-offline-save','passed':True});print(engine,'update/offline PASS',flush=True)
finally:
 network=True;server.shutdown();(OUT/'results.json').write_text(json.dumps(results,indent=2))

assert all(item['passed'] for item in results), 'Browser cases failed; see results.json'

File: tests/browser.mjs
import assert from 'node:assert/strict';
import {createServer} from 'node:http';
import {readFile,mkdir,writeFile} from 'node:fs/promises';
import {resolve,extname} from 'node:path';
import {chromium,webkit} from 'playwright';

// Serve under the same subdirectory used by the GitHub Pages project.
const root=resolve('.'),prefix='/General-AI-Test-Space-5/';
let networkAvailable=true;
const mime={'.html':'text/html','.js':'text/javascript','.css':'text/css','.svg':'image/svg+xml','.png':'image/png','.webmanifest':'application/manifest+json'};
const server=createServer(async(req,res)=>{if(!networkAvailable){res.destroy();return;}try{const path=decodeURIComponent(new URL(req.url,'http://localhost').pathname);if(!path.startsWith(prefix)){res.writeHead(404).end();return;}const file=resolve(root,path.slice(prefix.length)||'index.html');if(!file.startsWith(root+'/')){res.writeHead(403).end();return;}const data=await readFile(file);res.writeHead(200,{'Content-Type':mime[extname(file)]||'text/plain','Cache-Control':'no-cache'});res.end(data);}catch{res.writeHead(404).end();}});
await new Promise(done=>server.listen(8765,'127.0.0.1',done));
await mkdir('qa',{recursive:true});
const results=[];
const profiles=[{name:'desktop',viewport:{width:1365,height:900},touch:false},{name:'iphone-portrait',viewport:{width:402,height:874},touch:true},{name:'iphone-landscape',viewport:{width:874,height:402},touch:true}];
for(const [engine,type] of [['chromium',chromium],['webkit',webkit]]){
 const browser=await type.launch({headless:true});
 for(const profile of profiles){
  networkAvailable=true;
  const name=engine+'-'+profile.name,checks=[],errors=[];
  const context=await browser.newContext({viewport:profile.viewport,isMobile:profile.touch,hasTouch:profile.touch,deviceScaleFactor:1});
  const page=await context.newPage();page.setDefaultTimeout(15000);page.on('pageerror',e=>errors.push(e.message));
  const check=(name,value)=>{assert.ok(value,name);checks.push(name);};
  const shot=label=>page.screenshot({path:`qa/${name}-${label}.png`,fullPage:true});
  const state=()=>page.evaluate(()=>__deadblock.game?.snapshot());
  const press=async selector=>{const locator=page.locator(selector).last();if(profile.touch)await locator.tap();else await locator.click();};
  try{
   await page.goto(`http://127.0.0.1:8765${prefix}`,{waitUntil:'networkidle'});
   await page.waitForFunction(()=>window.__deadblock&&document.querySelector('#bootStatus').hidden);
   await page.waitForFunction(()=>document.querySelector('#offlineStatus').textContent.includes('OFFLINE READY'));
   console.log('CAPABILITIES',name,await page.evaluate(()=>({maxTouchPoints:navigator.maxTouchPoints,coarse:matchMedia('(pointer:coarse)').matches,setting:__deadblock.settings.touch})));
   check('lobby loads; no horizontal overflow',await page.evaluate(()=>document.documentElement.scrollWidth<=innerWidth));
   check('all 18 rooms available',await page.locator('#mapSelect option').count()===18);
   for(const id of await page.locator('#mapSelect option').evaluateAll(a=>a.map(e=>e.value)))await page.selectOption('#mapSelect',id);
   await page.selectOption('#mapSelect','outpost');await shot('lobby');
   await press('#startButton');await page.waitForTimeout(350);
   check('new game creates one save',await page.evaluate(()=>!!__deadblock.store.load()&&__deadblock.view==='play'));
   if(profile.touch)check('touch input automatically displays joysticks',await page.locator('#movePad').isVisible());
   await press('#pauseButton');const frozen=(await state()).time;await page.waitForTimeout(200);
   check('pause freezes simulation',(await state()).time===frozen);await shot('pause');
   await press('[data-action="resume"]');
   const before=(await state()).player;
   if(profile.touch){const pad=await page.locator('#movePad').boundingBox();assert.ok(pad,'movement pad visible');await page.mouse.move(pad.x+pad.width/2,pad.y+pad.height/2);await page.mouse.down();await page.mouse.move(pad.x+pad.width/2+32,pad.y+pad.height/2-15);await page.waitForTimeout(250);await page.mouse.up();}
   else{await page.keyboard.down('KeyD');await page.waitForTimeout(250);await page.keyboard.up('KeyD');}
   check('movement input changes position',(await state()).player.x>before.x+10);
   if(profile.touch&&engine==='chromium'){
    const left=await page.locator('#movePad').boundingBox(),right=await page.locator('#aimPad').boundingBox();const cdp=await context.newCDPSession(page);
    const points=[{id:1,x:left.x+left.width/2,y:left.y+left.height/2},{id:2,x:right.x+right.width/2,y:right.y+right.height/2}];
    await cdp.send('Input.dispatchTouchEvent',{type:'touchStart',touchPoints:points});points[0].x+=28;points[1].x+=30;points[1].y-=10;await cdp.send('Input.dispatchTouchEvent',{type:'touchMove',touchPoints:points});await page.waitForTimeout(180);
    check('simultaneous touch move and aim',await page.evaluate(()=>__deadblock.input.move.x>.2&&__deadblock.input.aim.x>.2));
    await cdp.send('Input.dispatchTouchEvent',{type:'touchEnd',touchPoints:[]});await page.waitForTimeout(30);
    check('touch release resets both sticks',await page.evaluate(()=>__deadblock.input.move.x===0&&__deadblock.input.aim.x===0));await cdp.detach();
   }
   await page.waitForTimeout(1600);await shot('gameplay');
   await page.evaluate(()=>{const g=__deadblock.game;g.s.spawnLeft=0;g.s.enemies=[];g.s.projectiles=[];g.s.cash=10000;g.update(1/60,{});__deadblock.processEvents();});
   await press('[data-action="buy"][data-id="uzi"]');
   check('armory purchase persists',await page.evaluate(()=>__deadblock.store.load().state.owned.includes('uzi')));
   await press('[data-tab="upgrades"]');await press('[data-action="upgrade"][data-id="damage"]');await press('[data-action="upgrade"][data-id="health"]');
   check('upgrade purchase persists',await page.evaluate(()=>__deadblock.store.load().state.up.damage===1&&__deadblock.game.maxHp===120));
   await press('[data-tab="weapons"]');await shot('armory');const saved=await state();
   await page.reload({waitUntil:'networkidle'});await press('#continueButton');
   check('reload restores exact run state',JSON.stringify(await state())===JSON.stringify(saved));
   await press('[data-action="next"]');check('next wave advances',(await state()).wave===2);
   await page.evaluate(()=>window.dispatchEvent(new Event('pagehide')));const backgroundTime=(await state()).time;await page.waitForTimeout(180);
   check('leaving app pauses without background combat',(await state()).time===backgroundTime&&await page.evaluate(()=>__deadblock.paused));
   await press('[data-action="resume"]');check('resume reacquires save ownership',await page.evaluate(()=>!__deadblock.paused&&__deadblock.store.owns(__deadblock.game.s.runId)));
   await press('#pauseButton');await press('[data-action="menu"]');const oldId=await page.evaluate(()=>__deadblock.store.load().state.runId);
   await press('#startButton');await press('[data-action="cancel-new"]');check('cancel new game keeps save',await page.evaluate(()=>__deadblock.store.load().state.runId)===oldId);
   await press('#startButton');await press('[data-action="confirm-new"]');
   check('confirmed new game replaces the one slot',(await state()).runId!==oldId&&(await state()).owned.length===1&&(await state()).up.damage===0);
   await page.evaluate(()=>{const g=__deadblock.game;g.s.score=12345;g.s.player.hurt=0;g.hurt(10000);__deadblock.processEvents();});
   check('death deletes save and keeps personal best',await page.evaluate(()=>__deadblock.store.load()===null&&__deadblock.store.records().score>=12345));await shot('death');
   await press('[data-action="menu"]');await page.reload({waitUntil:'networkidle'});
   check('dead run does not return after reload',await page.locator('#continueButton').isHidden());
   // A socket-level outage tests cached navigation in both engines. WebKit's
   // DevTools offline flag can fail internally before consulting its worker.
   networkAvailable=false;
   await page.reload({waitUntil:'load'});await page.waitForFunction(()=>window.__deadblock&&document.querySelector('#startButton').disabled===false);await press('#startButton');
   check('cached app starts with its server unreachable',await page.evaluate(()=>__deadblock.view==='play'&&__deadblock.game.s.phase==='wave'));
   networkAvailable=true;
   check('no uncaught browser errors',errors.length===0);
   results.push({name,passed:true,checks,errors});console.log('PASS',name,checks.length,'checks');
  }catch(error){await shot('FAILED').catch(()=>{});results.push({name,passed:false,checks,errors,error:String(error.stack||error)});console.error('FAIL',name,error);}
  networkAvailable=true;await context.close();
 }
 await browser.close();
}
server.close();await writeFile('qa/results.json',JSON.stringify(results,null,2));
console.log(JSON.stringify(results.map(r=>({name:r.name,passed:r.passed,checks:r.checks.length})),null,2));
if(results.some(r=>!r.passed))process.exitCode=1;
